### ISSUE TITLE
Sync issues on v3.0

### ISSUE THREAD

**Issue by pentool** [HappenApps/Quiver#287]
Hi,

I seem to have constant syncing issues. My desktop and MacBook does not seem to sync. They both setup for iCloud Drive syncing. I've noticed that two computers can fall out of sync quite fast. I am not sure if this is an iCloud Drive issue or a Quiver issue.

I use a desktop Mac and a MacBook. Both computers running.
1. Open Quiver on the desktop and add/modify a notebook/item.
2. Close Quiver.
3. Open Quiver on the MacBook and add/modify a notebook/item.

I am not sure if there's a sync interval between Quiver an iCloud Drive, but the entries I've added on the desktop are not showing up on the MacBook.
1. Wait a while.
2. Open Quiver on the MacBook.
3. If you are lucky, after a waiting period, the changes you made on the desktop will be synced to the MacBook. But for some reason, it seems this does not happen at times. Not sure if this is an iCloud Drive issue.

However, it seems the Trash items are not syncing either. This gets interesting when trashed items have tags, and the tag is used only by the item that is in the trash.
1. Create an item on the desktop and assign a tag to it.
2. Do whatever is necessary (eg wait, etc) until this item along with the tag syncs to the MacBook as well.
3. Close Quiver on the desktop.
4. Delete the new item on the MacBook. This will simply move the item to the Trash and the tag still on the Tag list.
5. Empty the Trash. This gets rid of the item in the trash.
6. Close Quiver on the MacBook.
7. Open Quiver on the desktop and quickly delete the item you created in step #1 above. This will simply move the item to the Trash. DO NOT empty the trash!
8. Close Quiver.
9. Open Quiver on the MacBook. The item you placed to the Trash on the desktop is not synced. This is probably intentional, but now notice the tag you crated in step #1 is not showing on the MacBook, but it does showing on the desktop!
10. Close Quiver on the MacBook.
11. Recover the item from the Trash on the desktop.

At this point, everything just goes downhill. Now you have restored the item from the Trash on the desktop (and it has a tag assigned). On the MacBook, however, this item will not sync. So now you are 1 item and 1 tag behind!

You can do all sorts of variations on the above, to make sync somehow break. I've also tried to...
1. On the MacBook, go to Preferences > Sync and select "Open Another Library". This is in attempt to pull in the extra item with the tag that you have restored from the Trash on the desktop machine.
2. Now you get a dialog box saying that the modification are not in sync and you have to chose which one to keep.

Here's the problem. If you are not using Quiver for a few days (or weeks), you forget what have you done the last time and on which machine - despite the dates in this dialog box. Eventually it leads to problems down the road that your library will fall out of sync big time and you will have to keep figuring out which is the latest one.

I also have the app called "SnippetsLab" which is very similar to Quiver. That app using iCloud sync - as opposed to iCloud Drive sync. Not really sure what the difference is, but when you go to the iCloud System Preferences, you can see "SnippetsLab" showing up on the list. However, because Quiver is syncing via iCloud Drive, somehow that's a different sync method and so Quiver does not showing up under the iCloud System Preferences list.

I've tried the same game with SnippetsLab, but that does not seem to exhibit the above mentioned issues. So it seems that there are indeed differences between the two sync methods. Using the iCloud sync (that is used by SnippetsLab), the changes are close to immediate on both machines, so my notes are in sync. In Quiver, however, they can fall out of sync fast - possibly because of the iCloud Drive sync method.

(...sorry for the long post...)

**Comment by ylian** [comment]
Thanks. A few points:
1. iCloud Drive is different from iCloud. Quiver doesn't use iCloud internally, but only use iCloud Drive the same way as any other cloud file services. Essentially Quiver doesn't care how the files are synced, but treat them as local files.
2. iCloud Drive takes a while to sync. You can see the spinning icon on the iCloud Drive folder while it syncs. Sometimes it takes a few minutes.
3. There might be UI inconsistencies regarding the tag issue, but it shouldn't cause permanent "out of sync" issues. Quiver always check the local files on startup, so any "out of sync" issue should be addressed by a simple app restart. Of course I will check the UI inconsistencies during syncing.

**Comment by pentool** [comment]
Another weird issue I just encountered is that somehow, I ended up with an item that does not exist unless I select the only tag it is assigned to it.

Eg, I have a note titled "Another extra note" and this note has a tag assigned to it called "extra".
1. So I select the "All Notes" item in the Sidebar to display all my notes. NONE of the notes listed have the item titled "Another extra note". I even checked every single notebook I have. None of them has the note.
2. I switch to the "Tags" tab on the sidebar and select the "extra" tag. This tag shows that it has a single item assigned to it.
3. Select the "extra" tag. Sure enough the item "Another extra note" displays.

So it's weird that this note only appears if I select its tag, but otherwise it doesn't appear anywhere else. Not even in the trash (which is empty).

This somehow showed up while I was playing with the sync issue.

**Comment by ylian** [comment]
This sounds more like a bug with UI out of sync, instead of permanent data out of sync.

If you restart the app, is the extra note gone?

**Comment by JMichaelTX** [comment]
FWIW, **I've never had a problem with sync using DropBox**, other than an occasional popup asking if I wanted to keep changes I had just made.  I always click "Yes", and the sync has always been correct.

The other advantage of Dropbox is that is has version history that is easy to use.

**Comment by ylian** [comment]
There are two things here:
1. Permanent out of sync. This would be a serious problem, but I haven't seen this happen anywhere and no one else has reported this. Quiver always check the local files during startup, so I'm highly confident this doesn't happen. But maybe iCloud Drive messed up the files on disk.
2. UI out of sync temporarily during syncing. I know this can happen in some edge cases. Quiver relies on file change events to update the UI, and sometimes file change events are fired out of order. I will keep improving handling of those edge cases.

**Comment by pentool** [comment]
Thanks for the feedback! No, the first time I restarted Quiver the tag (and the item) did not go away. But after that I just manually deleted it.

Yesterday, I did noticed however, that there was a little lightning bolt (kinda) symbol inside my iCloud Drive icon (...maybe it was charging...? :-) I've never seen that before. But then again, I usually don't use iCloud Drive for anything (or any cloud service for that matter).

I will keep playing around with these more and will report back if I find anything odd (or more info).

**Comment by ylian** [comment]
Yeah that's the "sync" icon. That means iCloud Drive is still syncing. That's the problem. iCloud Drive often takes a long time (can be minutes) to sync changes, while Dropbox takes seconds consistently.

Another thing I do want to mention is that Quiver also cache notebook and tag data for performance reasons, so sometimes an empty tag might still show up after restarting the app, but it will certainly go away after restarting again. In general the idea is that things may go wrong for whatever reasons, but Quiver will reliably recover without the user manually fixing.

**Comment by pentool** [comment]
I see. Thanks, now it all start to make sense why things happen. I had a few occasions where before leaving for work in the morning, I was toying around on my desktop and entered some snippets, thinking that I will continue messing around with it at work. So I entered the snippet. Done. Shut down.

Taking the MacBook to work. Boot. Open Quiver. Where is my snippet!? Well, it didn't sync yet apparently cuz syncing was slow, and I shut down the desktop at home.

So now, I dug up the info again from the net that I entered at home before leaving, and entered it again in Quiver. At work. On the MacBook.

But now Quiver is syncing from the MacBook. When I get home and boot the desktop, it (apparently) continues the unfinished sync. So somehow I end up with 2 sets of libraries. And I guess this is why I get the dialog box (Preferences > Sync > Open Another Library) that Quiver is out of sync and I should pick a library to use. And one library is from the MacBook and the other is from the desktop.

I guess now I generated a wishlist for iCloud syncing :-)
